tests/integration/diag_scripts/mlr/_sklearn_utils.py

```diff
@@ -0,0 +1,295 @@
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""
+
+import warnings
+
+import numpy as np
+import pytest
+import scipy as sp
+from numpy.testing import assert_allclose, assert_array_equal
+from sklearn.base import BaseEstimator
+
+
+class FailingClassifier(BaseEstimator):
+    """Classifier that deliberately fails when using fit()."""
+
+    FAILING_PARAMETER = 42
+
+    def __init__(self, parameter=None):
+        """Initialize class instance."""
+        self.parameter = parameter
+
+    def fit(self, *_, **__):
+        """Fit."""
+        if self.parameter == FailingClassifier.FAILING_PARAMETER:
+            raise ValueError("Failing classifier failed as required")
+
+    @staticmethod
+    def predict(x_data):
+        """Predict."""
+        return np.zeros(x_data.shape[0])
+
+    @staticmethod
+    def score(*_, **__):
+        """Score."""
+        return 0.0
+
+
+def assert_warns(warning_class, func, *args, **kw):
+    """Test that a certain warning occurs.
+
+    Parameters
+    ----------
+    warning_class : the warning class
+        The class to test for, e.g. UserWarning.
+
+    func : callable
+        Callable object to trigger warnings.
+
+    *args : the positional arguments to `func`.
+
+    **kw : the keyword arguments to `func`
+
+    Returns
+    -------
+    result : the return value of `func`
+
+    """
+    with warnings.catch_warnings(record=True) as warn:
+        # Cause all warnings to always be triggered.
+        warnings.simplefilter("always")
+        # Trigger a warning.
+        result = func(*args, **kw)
+        if hasattr(np, 'FutureWarning'):
+            # Filter out numpy-specific warnings in numpy >= 1.9
+            warn = [e for e in warn
+                    if e.category is not np.VisibleDeprecationWarning]
+
+        # Verify some things
+        if not len(warn) > 0:
+            raise AssertionError("No warning raised when calling %s"
+                                 % func.__name__)
+
+        found = any(warning.category is warning_class for warning in warn)
+        if not found:
+            raise AssertionError("%s did not give warning: %s( is %s)"
+                                 % (func.__name__, warning_class, warn))
+    return result
```

Comment:
this looks like a function that may sit better in the tests `__init__` don't you think?

Reply:
I'd like to keep this in this module, since this is again copy-pasted from `sklearn`, so it should stay here together with the copyright information.

Reply:
fine in that case :+1: